File: packages/core/src/modules/narrow-waist-bypass.ts
/**
 * R5 — Install-Orchestrator Bypass Detection
 * Detects direct access to keychain APIs, audit log writes,
 * mcp_config.json writes, and capability-registry.json writes
 * outside the allowed orchestrator modules.
 *
 * Aligned with Semgrep YAML v0.1.2 rules:
 * - quickwork-R5-js-direct-keychain-access
 * - quickwork-R5-py-direct-keychain-access
 * - quickwork-R5-direct-audit-log-write
 * - quickwork-R5-direct-quick-config-write
 * - quickwork-R5-direct-capability-registry-write
 */

import * as fs from 'fs';
import * as path from 'path';
import { ScanContext, ScanFinding, ScannerModule } from '../types';

interface BypassPattern {
  name: string;
  regex: RegExp;
  message: string;
  ruleId: string;
  category: string;
  recommendation: string;
}

const JS_BYPASS_PATTERNS: BypassPattern[] = [
  // Keychain / credential APIs
  {
    name: 'keytar-import',
    regex: /require\(['"]keytar['"]\)/g,
    message: 'Direct keytar import (bypass orchestrator credentials API)',
    ruleId: 'R5-direct-keychain-access',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/credentials/* API instead of keytar',
  },
  {
    name: 'node-credential-manager-import',
    regex: /require\(['"]node-credential-manager['"]\)/g,
    message: 'Direct node-credential-manager import (bypass orchestrator)',
    ruleId: 'R5-direct-keychain-access',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/credentials/* API',
  },
  {
    name: 'credential-manager-import',
    regex: /require\(['"]credential-manager['"]\)/g,
    message: 'Direct credential-manager import (bypass orchestrator)',
    ruleId: 'R5-direct-keychain-access',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/credentials/* API',
  },
  {
    name: 'macos-security-command',
    regex: /spawn\(['"]security['"]|exec\(['"]security\s+add-generic-password/g,
    message: 'Direct macOS security command (bypass orchestrator credentials API)',
    ruleId: 'R5-direct-keychain-access',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/credentials/* API',
  },
  // Audit log direct write
  {
    name: 'audit-log-direct-write',
    regex: /writeFile(?:Sync)?\(['"`](?:~\/\.quickwork\/|.*?quickport\/state\/)\.audit\.json['"`]/g,
    message: 'Direct audit log write (bypass orchestrator audit API)',
    ruleId: 'R5-direct-audit-log-write',
    category: 'privilege-escalation',
    recommendation: 'Use official audit logging APIs',
  },
  // mcp_config.json direct write (string literals)
  {
    name: 'mcp-config-string-literal',
    regex: /writeFile(?:Sync)?\(['"`][^'"`]*mcp_config\.json['"`]/g,
    message: 'Direct mcp_config.json write (bypass quick-config-patcher 5-invariants protocol)',
    ruleId: 'R5-direct-quick-config-write',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/quick-config-patcher/* API with atomic write + backup + schema validate + audit + rollback',
  },
  // mcp_config.json template literals
  {
    name: 'mcp-config-template-literal',
    regex: /writeFile(?:Sync)?\(`\$\{[^}]*\}\/\.quickwork\/mcp_config\.json`/g,
    message: 'Dynamic mcp_config.json path write (bypass quick-config-patcher)',
    ruleId: 'R5-direct-quick-config-write',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/quick-config-patcher/* API',
  },
  // mcp_config.json path.join
  {
    name: 'mcp-config-path-join',
    regex: /writeFile(?:Sync)?\(path\.join\([^)]*['"`].quickwork['"`],?\s*['"`]mcp_config\.json['"`]\)/g,
    message: 'path.join mcp_config.json write (bypass quick-config-patcher)',
    ruleId: 'R5-direct-quick-config-write',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/quick-config-patcher/* API',
  },
  // capability-registry.json direct write
  {
    name: 'capability-registry-string-literal',
    regex: /writeFile(?:Sync)?\(['"`][^'"`]*capability-registry\.json['"`]/g,
    message: 'Direct capability-registry.json write (bypass capability-registry module)',
    ruleId: 'R5-direct-capability-registry-write',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/capability-registry/* API',
  },
  // capability-registry.json template literals
  {
    name: 'capability-registry-template-literal',
    regex: /writeFile(?:Sync)?\(`\$\{[^}]*\}\/quickport\/state\/capability-registry\.json`/g,
    message: 'Dynamic capability-registry.json path write (bypass capability-registry)',
    ruleId: 'R5-direct-capability-registry-write',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/capability-registry/* API',
  },
];

const PY_BYPASS_PATTERNS: BypassPattern[] = [
  {
    name: 'python-keyring-access',
    regex: /keyring\.(set_password|get_password|delete_password)\s*\(/g,
    message: 'Direct Python keyring access (bypass orchestrator credentials API)',
    ruleId: 'R5-py-direct-keychain-access',
    category: 'privilege-escalation',
    recommendation: 'Use @quickport/orchestrator/credentials/* API',
  },
];

function checkFile(
  filePath: string,
  relPath: string,
  patterns: BypassPattern[],
  findings: ScanFinding[]
): void {
  const content = fs.readFileSync(filePath, 'utf-8');

  for (const pattern of patterns) {
    const matches = content.match(pattern.regex);
    if (matches) {
      for (const match of matches) {
        findings.push({
          ruleId: pattern.ruleId,
          tier: 'blocker',
          severity: 'P0',
          criticalTag: '[critical:security]',
          message: pattern.message,
          file: relPath,
          category: pattern.category as any,
          evidence: match.substring(0, 80),
          recommendation: pattern.recommendation,
          ruleOrigin: 'core',
        });
      }
    }
  }
}

export class NarrowWaistBypassModule implements ScannerModule {
  name = 'narrow-waist-bypass';

  scan(ctx: ScanContext): ScanFinding[] {
    const findings: ScanFinding[] = [];

    const jsFiles = ctx.sourceFiles.filter((f) =>
      /\.(js|ts|jsx|tsx|mjs|cjs)$/.test(f)
    );

    const pyFiles = ctx.sourceFiles.filter((f) => f.endsWith('.py'));

    for (const relPath of jsFiles) {
      // Skip orchestrator modules (they are allowed to access these APIs)
      if (relPath.includes('orchestrator/')) continue;
      // Skip test files (false positive exclusion per Semgrep YAML)
      if (/\.(test|spec)\.(ts|js)$/.test(relPath)) continue;
      if (relPath.includes('/test/') || relPath.includes('/tests/')) continue;
      if (relPath.includes('/fixtures/')) continue;

      const fullPath = path.join(ctx.skillPath, relPath);
      checkFile(fullPath, relPath, JS_BYPASS_PATTERNS, findings);
    }

    for (const relPath of pyFiles) {
      if (relPath.includes('orchestrator/')) continue;
      if (/\.(test|spec)\.py$/.test(relPath)) continue;
      if (relPath.includes('/test/') || relPath.includes('/tests/')) continue;

      const fullPath = path.join(ctx.skillPath, relPath);
      checkFile(fullPath, relPath, PY_BYPASS_PATTERNS, findings);
    }

    return findings;
  }
}

File: crates/core/src/modules/narrow_waist_bypass.rs
use std::fs;
use std::path::Path;

use regex::Regex;

use crate::types::{
    CriticalTag, RuleOrigin, ScanContext, ScanFinding, ScannerModule, Severity, ThreatCategory,
    Tier,
};

struct BypassPattern {
    regex: Regex,
    message: &'static str,
    rule_id: &'static str,
    recommendation: &'static str,
}

fn js_bypass_patterns() -> Vec<BypassPattern> {
    vec![
        BypassPattern {
            regex: Regex::new(r#"require\(['"]keytar['"]\)"#).unwrap(),
            message: "Direct keytar import (bypass orchestrator credentials API)",
            rule_id: "R5-direct-keychain-access",
            recommendation: "Use @quickport/orchestrator/credentials/* API instead of keytar",
        },
        BypassPattern {
            regex: Regex::new(r#"require\(['"]node-credential-manager['"]\)"#).unwrap(),
            message: "Direct node-credential-manager import (bypass orchestrator)",
            rule_id: "R5-direct-keychain-access",
            recommendation: "Use @quickport/orchestrator/credentials/* API",
        },
        BypassPattern {
            regex: Regex::new(r#"require\(['"]credential-manager['"]\)"#).unwrap(),
            message: "Direct credential-manager import (bypass orchestrator)",
            rule_id: "R5-direct-keychain-access",
            recommendation: "Use @quickport/orchestrator/credentials/* API",
        },
        BypassPattern {
            regex: Regex::new(r#"spawn\(['"]security['"]|exec\(['"]security\s+add-generic-password"#).unwrap(),
            message: "Direct macOS security command (bypass orchestrator credentials API)",
            rule_id: "R5-direct-keychain-access",
            recommendation: "Use @quickport/orchestrator/credentials/* API",
        },
        BypassPattern {
            regex: Regex::new(r#"writeFile(?:Sync)?\(['"`](?:~\/\.quickwork\/|.*?quickport\/state\/)\.audit\.json['"`]"#).unwrap(),
            message: "Direct audit log write (bypass orchestrator audit API)",
            rule_id: "R5-direct-audit-log-write",
            recommendation: "Use official audit logging APIs",
        },
        BypassPattern {
            regex: Regex::new(r#"writeFile(?:Sync)?\(['"`][^'"`]*mcp_config\.json['"`]"#).unwrap(),
            message: "Direct mcp_config.json write (bypass quick-config-patcher 5-invariants protocol)",
            rule_id: "R5-direct-quick-config-write",
            recommendation: "Use @quickport/orchestrator/quick-config-patcher/* API with atomic write + backup + schema validate + audit + rollback",
        },
        BypassPattern {
            regex: Regex::new(r#"writeFile(?:Sync)?\(['"`][^'"`]*capability-registry\.json['"`]"#).unwrap(),
            message: "Direct capability-registry.json write (bypass capability-registry module)",
            rule_id: "R5-direct-capability-registry-write",
            recommendation: "Use @quickport/orchestrator/capability-registry/* API",
        },
    ]
}

fn py_bypass_patterns() -> Vec<BypassPattern> {
    vec![BypassPattern {
        regex: Regex::new(r"keyring\.(set_password|get_password|delete_password)\s*\(").unwrap(),
        message: "Direct Python keyring access (bypass orchestrator credentials API)",
        rule_id: "R5-py-direct-keychain-access",
        recommendation: "Use @quickport/orchestrator/credentials/* API",
    }]
}

pub struct NarrowWaistBypassModule;

impl ScannerModule for NarrowWaistBypassModule {
    fn name(&self) -> &str {
        "narrow-waist-bypass"
    }

    fn scan(&self, ctx: &ScanContext) -> Vec<ScanFinding> {
        let mut findings = Vec::new();
        let js_ext_re = Regex::new(r"\.(js|ts|jsx|tsx|mjs|cjs)$").unwrap();
        let test_re = Regex::new(r"\.(test|spec)\.(ts|js)$").unwrap();

        let js_files: Vec<&String> = ctx
            .source_files
            .iter()
            .filter(|f| js_ext_re.is_match(f))
            .filter(|f| !test_re.is_match(f))
            .filter(|f| !f.contains("/test/") && !f.contains("/tests/"))
            .filter(|f| !f.contains("/fixtures/"))
            .filter(|f| !f.contains("orchestrator/"))
            .collect();

        let py_files: Vec<&String> = ctx
            .source_files
            .iter()
            .filter(|f| f.ends_with(".py"))
            .filter(|f| !f.contains("orchestrator/"))
            .filter(|f| !f.contains("/test/") && !f.contains("/tests/"))
            .collect();

        let js_patterns = js_bypass_patterns();
        let py_patterns_list = py_bypass_patterns();

        for rel_path in &js_files {
            let full_path = Path::new(&ctx.skill_path).join(rel_path);
            let content = match fs::read_to_string(&full_path) {
                Ok(c) => c,
                Err(_) => continue,
            };

            for pattern in &js_patterns {
                for m in pattern.regex.find_iter(&content) {
                    let evidence = &m.as_str()[..m.as_str().len().min(80)];
                    findings.push(ScanFinding {
                        rule_id: pattern.rule_id.to_string(),
                        tier: Tier::Blocker,
                        severity: Severity::P0,
                        critical_tag: Some(CriticalTag::Security),
                        message: pattern.message.to_string(),
                        file: Some((*rel_path).clone()),
                        line: None,
                        column: None,
                        category: ThreatCategory::PrivilegeEscalation,
                        evidence: Some(evidence.to_string()),
                        recommendation: Some(pattern.recommendation.to_string()),
                        rule_origin: Some(RuleOrigin::Core),
                        ref_anchor: None,
                        merged_from: None,
                    });
                }
            }
        }

        for rel_path in &py_files {
            let full_path = Path::new(&ctx.skill_path).join(rel_path);
            let content = match fs::read_to_string(&full_path) {
                Ok(c) => c,
                Err(_) => continue,
            };

            for pattern in &py_patterns_list {
                for m in pattern.regex.find_iter(&content) {
                    let evidence = &m.as_str()[..m.as_str().len().min(80)];
                    findings.push(ScanFinding {
                        rule_id: pattern.rule_id.to_string(),
                        tier: Tier::Blocker,
                        severity: Severity::P0,
                        critical_tag: Some(CriticalTag::Security),
                        message: pattern.message.to_string(),
                        file: Some((*rel_path).clone()),
                        line: None,
                        column: None,
                        category: ThreatCategory::PrivilegeEscalation,
                        evidence: Some(evidence.to_string()),
                        recommendation: Some(pattern.recommendation.to_string()),
                        rule_origin: Some(RuleOrigin::Core),
                        ref_anchor: None,
                        merged_from: None,
                    });
                }
            }
        }

        findings
    }
}
